EDU-3236: Update-with-Start (#3120)
temporalio/documentation: docs/encyclopedia/application-message-passing.mdx

- The Workflow's clients want to quickly move on after sending an asynchronous message.
- The clients care more about the latency of sending the message than they do about the latency of the end-to-end operation.
- The clients don't want to rely on the Worker being available.

> - Clients need to [Signal-With-Start](#signal-with-start) atomically (since Update-with-Start is not atomic).

Reviewer: I vote to cut this.  When I wrote it initially, I really meant "atomic" imprecisely as "one API call."
The number of users who specifically care about atomicity is probably small enough that this gotcha can be relegated to the more detailed look at UwS.

    <RelatedReadItem path="/develop/dotnet/message-passing#send-update-from-client" text="Send Updates in .NET" archetype="feature-guide" />
</RelatedReadContainer>

#### Update-With-Start {#update-with-start}

> Sending an Update-With-Start, if there is a running Workflow Execution with the given Workflow Id, it will process the Update. Otherwise, a new Workflow Execution starts and immediately processes the Update.

Reviewer 2: ```suggestion
Update-With-Start lets you send an Update that checks whether an already-running Workflow with that ID exists.
If it does, the Update is delivered.
If not, it starts a new Workflow Execution with the supplied ID.
Upon starting a new Workflow, it immediately processes the Update.
```

Update-With-Start is great for latency-sensitive use cases:

- **Lazy Initialization** - Instead of making separate Start Workflow and Update Workflow calls, Update-With-Start allows to send them together in a single roundtrip. For example, a shopping cart can be modelled using Update-With-Start: Adding or removing an item with an Update; ensuring the cart is created first before applying the Update.

> - **Early Return** - Starting a new Workflow Execution and synchronously receiving a response mid-Workflow, while the Workflow Execution continues to run to completion. For example, a payment process can be modelled using Update-With-Start: Sending the result of the payment validation back to the client synchronously, and continuing to complete or rollback the transaction asynchronously.

Reviewer 2: ```suggestion
- **Early Return** - 
  Using Update-With-Start can begin a new Workflow Execution and synchronously receiving a response while the Workflow Execution continues to run to completion.
  For example, a payment process can be modelled using Update-With-Start.
  You can send the result of payment validation back to the client synchronously, while continuing to complete or rollback the transaction asynchronously.
```

> It is important to note that - unlike Signal-with-Start - Update-With-Start is *not* atomic. That means if the Update cannot be delievered - for example, when no Worker is running - the Workflow Execution is still started. The SDKs will retry the Update-With-Start request; but there is no gurantee that it will succeed.

Reviewer 2: ```suggestion

:::caution

Unlike Signal-with-Start - Update-With-Start is *not* atomic.
If the Update cannot be delievered, for example, when no Worker is running, the Workflow Execution is still started.
The SDKs will retry the Update-With-Start request but there is no guarantee that it will succeed.

:::
```

    <RelatedReadItem path="/develop/dotnet/message-passing#send-update-from-client" text="Send Updates in .NET" archetype="feature-guide" />
</RelatedReadContainer>

#### Update-With-Start {#update-with-start}

Update-With-Start lets you send an Update that checks whether an already-running Workflow with that ID exists.

> If it does, the Update is delivered.

Reviewer: nit: is processed ?  (playing up the synchronous nature)

If not, it starts a new Workflow Execution with the supplied ID.
Upon starting a new Workflow, it immediately processes the Update.

Update-With-Start is great for latency-sensitive use cases, including the following examples:

- **Lazy Initialization** -
  Instead of making separate Start Workflow and Update Workflow calls, Update-With-Start allows you to send them together in a single roundtrip.
  For example, a shopping cart can be modelled using Update-With-Start.
  Add or remove an item with an Update.

>   Update-With-Start ensures the cart is created first (the Workflow Execution) before applying the Update (state change of items within the cart).

Reviewer: ```suggestion
  Update-With-Start ensures the cart (the Workflow Execution) is created first before applying the Update (state change of items within the cart).
```

- **Early Return** -
  Using Update-With-Start can begin a new Workflow Execution and synchronously receiving a response while the Workflow Execution continues to run to completion.
  For example, a payment process can be modelled using Update-With-Start.
  You can send the result of payment validation back to the client synchronously, while continuing to complete or rollback the transaction asynchronously.

:::caution

Unlike Signal-with-Start - Update-With-Start is *not* atomic.

> If the Update cannot be delievered, for example, when no Worker is running, the Workflow Execution is still started.

Reviewer: ```suggestion
If the Update cannot be delivered, for example, when no Worker is running, the Workflow Execution is still started.
```

    <RelatedReadItem path="/develop/dotnet/message-passing#send-update-from-client" text="Send Updates in .NET" archetype="feature-guide" />
</RelatedReadContainer>

#### Update-With-Start {#update-with-start}

:::tip Stability

In [Pre-release](/evaluate/development-production-features/release-stages#pre-release) (API is subject to change).

Minimum recommended [Temporal Server version 1.25](https://github.com/temporalio/temporal/releases/tag/v1.25.0)

:::

> Update-With-Start lets you send an Update that checks whether an already-running Workflow with that ID exists.

Reviewer 2: ```suggestion
Update-With-Start sends an Update that checks whether an already-running Workflow with that ID exists.
```

If it does, the Update is processed.
If not, it starts a new Workflow Execution with the supplied ID.

> Upon starting a new Workflow, it immediately processes the Update.

Reviewer 2: ```suggestion
When starting a new Workflow, it immediately processes the Update.
```

> Update-With-Start is great for latency-sensitive use cases, including the following examples:

Reviewer 2: ```suggestion
Update-With-Start is great for latency-sensitive use cases:
```

- **Lazy Initialization** -
  Instead of making separate Start Workflow and Update Workflow calls, Update-With-Start allows you to send them together in a single roundtrip.
  For example, a shopping cart can be modelled using Update-With-Start.

>   Add or remove an item with an Update.

Reviewer 2: ```suggestion
Updates let you add and remove items from the cart.
```

>   Update-With-Start ensures the cart (the Workflow Execution) is created first before applying the Update (state change of items within the cart).

Reviewer 2: ```suggestion
  Update-With-Start lets the customer start shopping, whether the cart already exists or they've just started shopping.
  It ensures the cart, the Workflow Execution, is created before applying any Update that changes the state of items within the cart.
```

- **Early Return** -
  Using Update-With-Start can begin a new Workflow Execution and synchronously receiving a response while the Workflow Execution continues to run to completion.

>   For example, a payment process can be modelled using Update-With-Start.

Reviewer 2: ```suggestion
  For example, you might model a payment process using Update-With-Start.
```

>   You can send the result of payment validation back to the client synchronously, while continuing to complete or rollback the transaction asynchronously.

Reviewer 2: ```suggestion
  Whether the payment completes or rolls back the transaction, you can send the payment validation results back to the client synchronously, while attempting the transaction asynchronously.
```
Close?

> :::caution

Reviewer 2: I love this admonition.

Unlike Signal-with-Start - Update-With-Start is *not* atomic.

> If the Update cannot be delivered, for example, when no Worker is running, the Workflow Execution is still started.

Reviewer 2: ```suggestion
If the Update can't be delivered, for example, say that there's no running Worker available, a new Workflow Execution will still start.
```
